BCStudentSoftwareDevTeam/celts
# 1133 bonner cohort control

# Add New Cohorts to Bonner

## Issue

Resolves #1133  
Currently, a new cohort year is added automatically at a certain date, we should change it to have a button to manually add more cohort years instead of having it happen automatically.

This allows administrators to add people to a cohort by hand as early as they know who should be in it.

## Solution

We added a button that adds a new cohort to the Bonner Management page.
![image](https://github.com/BCStudentSoftwareDevTeam/celts/assets/89316750/162ecc77-7b2c-4c6e-a09a-4d3302bd82f2)
This button does not permanently add a cohort but rather makes it visible for the current session so that administrators can add students to it which then makes the cohort persistent.

## To Test

### Setup

From the `celts` directory
- `git checkout 1133-Bonnor-Cohort-Control` (yes, I know I spelled Bonner incorrectly in the branch title but in my defense, that's how it should be spelled)
- `database/reset_database.sh from-backup`
- `flask run`
- Navigate to Bonner Management using the side menu
![image](https://github.com/BCStudentSoftwareDevTeam/celts/assets/89316750/c39b8761-c98c-47c5-99d8-1f21a3135d9f)

### Testing

- Add users to a cohort of your choice and refresh the page, ensuring that they persist and no functionality is lost.
- Add a few new cohorts and add students to some of them ensuring that the cohorts are displayed on page load when they have non-zero student counts.
- Remove everyone from the most recently added cohort (furthest in the future) and the page will reload automatically. That cohort should disappear. Ensure that all empty cohorts after the current year are not displayed.
- Ensure that this system feels natural to use and isn't jarring when cohorts are removed on page reload if they're empty.
